### ISSUE TITLE
rollup-plugin-posthtml

### ISSUE THREAD

**Issue by Shyam-Chen** [posthtml/posthtml#199]
rollup-plugin-posthtml

**Comment by michael-ciniawsky** [comment]
@Shyam-Chen You mean `rollup-plugin-posthtml` or to provide a bundled version of the source?

First could be derivated from [rollup-plugin-postcss](https://github.com/egoist/rollup-plugin-postcss) :)

**Comment by Shyam-Chen** [comment]
Yes, I mean `rollup-plugin-posthtml`.

**Comment by michael-ciniawsky** [comment]
I can try making one, can you describe your usecase a bit futher please, I'm not too familiar with rollup + markup bundled with it.

**Comment by Shyam-Chen** [comment]
like Pug's Includes, Inheritance...

**Comment by tunnckoCore** [comment]
@Shyam-Chen it could be better to directly pass options to both `posthtml` and `.process`.

```js
      return posthtml(options.plugins)
        .process(code, options)
        .then((result) => ({ code: result.html, map: result.map }));
```

**Comment by Shyam-Chen** [comment]
I use `webpack`, `rollup-loader` and `posthtml-loader` to handle it.

**Comment by Shyam-Chen** [comment]
Can anyone help me?

https://github.com/Vanilla-IceCream/rollup-plugin-posthtml

**Comment by tunnckoCore** [comment]
whats the problem?

**Comment by michael-ciniawsky** [comment]
@Shyam-Chen I've seen it a while ago and can remember something small was bugging me, but I need to take a look again tbh 😛 I try to do it later. The plugin is 👍

**Comment by Shyam-Chen** [comment]
@tunnckoCore I want to import HTML as string.

**Comment by tunnckoCore** [comment]
@Shyam-Chen I don't see problem in current master implementation of the plugin. 

That seems correct `export default ${JSON.stringify(code)}`

**Comment by michael-ciniawsky** [comment]
@Shyam-Chen 

```diff
import { createFilter } from 'rollup-pluginutils';
import posthtml from 'posthtml';

export default function(options = {}) {
  const filter = createFilter(options.include, options.exclude);

  return {
    name: 'posthtml',
    transform(code, id) {
      if (!filter(id)) return;

      return posthtml(options.plugins || [])
        .process(code, {
          parser: options.parser,
          // options.from, options.to, options.render are aviable
-          template: options.template || false // Doesn't existst, see below for solution
        })
        .then(result => {
          let code, map;

-         code = `export default ${JSON.stringify(result.html)};`;
+         code = options.template 
+             ? export default `(_) => ${result.html}` // '_'  is just a random parameter name
+             : export default `${JSON.stringify(result.html)}
-          map = result.map; // Doesn't exists and Sourcemaps for HTML have no browser support

          return { code, map };
        });
    }
  };
}
```

**Comment by michael-ciniawsky** [comment]
For rollup plugin specifics you need to ask someone else 😛 .Interested in putting this plugin under posthtml org ? :)

**Comment by Shyam-Chen** [comment]
@michael-ciniawsky 

OK, let's put it in the PostHTML organization.

**Comment by michael-ciniawsky** [comment]
> OK, let's put it in the PostHTML organization

cc @voischev 

@Shyam-Chen Does it work now ? I'm not 💯  if the return `return html` `{String}` is working with the rollup API `return { html, map }` `{Object}` 😛

**Comment by tunnckoCore** [comment]
@michael-ciniawsky best selution is to return `{ code, map }` when not using `opts.template` and just `return code` would work when `opts.template` is used

**Comment by Shyam-Chen** [comment]
I modified the code, but my [example](https://github.com/Vanilla-IceCream/rollup-plugin-posthtml/tree/master/example) was still unable to work.

**Comment by michael-ciniawsky** [comment]
@tunnckoCore Are both return types `{String|Object}` supported by the Rollup API ?
@Shyam-Chen Can you `console.log()` and post the logs here :)

**Comment by tunnckoCore** [comment]
@michael-ciniawsky yes, even you can return a promise too :)

**Comment by tunnckoCore** [comment]
@Shyam-Chen about https://github.com/Vanilla-IceCream/rollup-plugin-posthtml/blob/master/example/rollup.config.js#L8-L9 - it's not needed, we are not in the crazy webpack land that don't handle such common and simple cases.

I think your problem is that the bundle is IIFE named `posthtml` by the moduleName, which actually isn't called, i mean.. not used.. mm not sure, i can't look too deeply now.

**Comment by tunnckoCore** [comment]
@Shyam-Chen doh, yea.. actually you must provide `options.include` (or include some default), because Rollup's filter does not have defaults. 

So yea, forget everything other that i said, set defaults for the `opts.include` to resolve html files and it will work.

**Comment by tunnckoCore** [comment]
@Shyam-Chen https://github.com/Vanilla-IceCream/rollup-plugin-posthtml/blob/master/src/index.js#L13 - just pass options, it would be better and enough.

**Comment by Shyam-Chen** [comment]
Sorry, I forgot to add this;

```js
if (!options.include) {
  options.include = '**/*.html'
}
```

Has been successful!

**Comment by tunnckoCore** [comment]
As about Line 12 there, no need for `options.plugins || []` since PostHTML already does this https://github.com/posthtml/posthtml/blob/master/lib/index.js#L29

**Comment by tunnckoCore** [comment]
> Has been successful!

Cool.

**Comment by Shyam-Chen** [comment]
See [here](https://github.com/Vanilla-IceCream/rollup-plugin-posthtml/blob/master/example/bundle.js).

**Comment by tunnckoCore** [comment]
What?

**Comment by Shyam-Chen** [comment]
```js
// ...
var foo = "<p>Foo</p>\r\n<p>Bar</p>\r\n\r\n";
// ...
```

**Comment by Shyam-Chen** [comment]
@michael-ciniawsky Do you have to publish the first version?

**Comment by tunnckoCore** [comment]
and what? it is not a problem, it is added to innerhtml, and the function i immediately invoked so it will be added to example div

**Comment by michael-ciniawsky** [comment]
@Shyam-Chen I have no Organisation rights here, but normally you will be added to the collabs team and have full access to posthtml/rollup-plugin-posthtml repo, while npm publishing is maintainers/owners responsibility, this could maybe be done by the recently open sourced npm organisation feature with a team there in the future, but for now that's an entirely different topic 😛 . Long story short, if you want to use it now, just publish it and if/when transfering and the organisational changes are done, simply release a semver patch :)

**Comment by tunnckoCore** [comment]
What's the point of that `options.template`? And why you would need to import a html as string _through_ the plugin options?

**Comment by tunnckoCore** [comment]
It's release and published https://github.com/tunnckoCore/rollup-plugin-posthtml. I just can't get the logic of `options.template` so it still isn't implemented, because I can't test it.

**Comment by michael-ciniawsky** [comment]
@tunnckoCore I don't use Rollup so iI can't verify if works that way but e.g 

**component.html**
```html
<div>${ _.hello }y<div>
```

**component.js**
```js
import component from './component.html'

document.body.innerHTML = component({ hello: 'Hello World!'})
```

Simple runtime interpolation via ES2015 template literals, works in `posthtml-loader`  during  build time use `posthtml-expressions` instead

**Comment by tunnckoCore** [comment]
How these snippets relate to the `options.template` topic which i can't undrestand?

As about the plugin that i published.. it works, yea

**Comment by michael-ciniawsky** [comment]
@tunnckoCore ? 🙃 `options.template` is a `{Boolean}` to export the processed HTML as `{String)` when `false` and as template function `{Function}`  when `true`

**Comment by Shyam-Chen** [comment]
I used it out.

https://github.com/Vanilla-IceCream/rollup-plugin-posthtml/blob/master/example/rollup.config.js#L14

**Comment by tunnckoCore** [comment]
Ooh, got it, interesting idea. I'll add it. But instead `_` it should be something more meaningful, maybe `state` or `context` or `ctx`?

**Comment by Shyam-Chen** [comment]
I think it should need a `posthtml-templates`.

**Comment by michael-ciniawsky** [comment]
@tunnckoCore In `posthtml-loader` it defaults to `_` when `options.template === true // {Boolean}`, but it's possible to pass a custom parameter name to `options.template` 

```js
options.template = 'your custom name here...' } // `{String}` 
```

 and inside the loader/plugin 
 
```js
...
const template = typeof options.template === 'boolean' ? '_' : options.template
...
`export default (${template}) => ${result.html}`
```

**Comment by michael-ciniawsky** [comment]
@Shyam-Chen hmm...I'm not sure about that :stuck_out_tongue:. For interpolation/templating during build time, there is `posthtml-expressions` already, which is more powerful then ES2015 template literals standalone, and since for interpolation/templating during runtime, the defacto standard is/will be ES2015 template literals, avoiding to use a third party solution is recommended imho, so you can easily reuse or share the generated components without any 'vendor' lock/boilerplate.

**Comment by Shyam-Chen** [comment]
@tunnckoCore add `options.template`?

**Comment by michael-ciniawsky** [comment]
What is in general going on with with the switch and two repo's now? 🙃 😛  Could you guys please agree on join efforts and share maintenance...

**Comment by Shyam-Chen** [comment]
How to do now?

**Comment by Shyam-Chen** [comment]
I found a bug. 

Use `posthtml-include` and `posthtml-sugarml` together ...

```pug
// foo.sml
p Foo
include(src='bar.sml')
```

```pug
// bar.sml
p Bar
```

Output:

```html
<p>Foo</p>
p Bar
```

**Comment by michael-ciniawsky** [comment]
Is it a problem with async?

@Shyam-Chen Nope :) the current version isn't even async :D I have a PR with an overhaul of the plugin ready [here](https://github.com/michael-ciniawsky/posthtml-include), which supports custom parser via `tree.options.parser` if `file.extname !== .html`  otherwise uses the default parser for HTML files, so both `.ssml/.html`  should work alongside in the same setup, blocked by #195 again, but I used in my setup 😛 

**components/component**
```pug
div(class=import) Hello
```

**index.ssml**
```pug
body
  import(src=components/component.ssml)
```

```js
// options.from ? dirname(options.from) : process.cwd()
posthtml([ include() ])
  .process('index.ssml, { parser: sugarml(), from: './index.ssml' })
  .then((result) => result.html)
```

**Comment by Shyam-Chen** [comment]
Do I need to publish the `rollup-plugin-posthtml-template`?

**Comment by tunnckoCore** [comment]
Such things not make any sense to be inside a the "posthtml". it is totally different thing and _should_ (must) use another rollup plugin for such things. I'm going to publish rollup-plugin-jstransformer soon which will automagically support ~200 engines.

> Do I need to publish 

Not make any sense. There's always hundreds of "template engines" which do such thing and [es6-template](https://github.com/tunnckoCore/es6-template) is one of them and is exactly working with ES6 tagged templates behind the scene, working everywhere where has support of course. The only difference is that you pass a normal string and it works as tagged string.

Anyway. I'm not agree to include such options and conditions. In most cases such conditions means "make another module, because current module will do two things instead of _'one thing well'_".

**Comment by michael-ciniawsky** [comment]
@Shyam-Chen @tunnckoCore ? What are you doing 😛 ? `rollup-plugin-posthtml`/ `rollup-plugin-posthtml-template``jstransformers` Why do we need all of that ? **One** plugin with `{String|Function}` Static (HTML) || Dynamic (Template Function) ideally matching with how `posthtml-loader` does it (UX Consistency) and as official PostHTML Plugin 🙃 ?

**Comment by tunnckoCore** [comment]
I don't understand you. There is just `rollup-plugin-posthtml`.

I still can't consider to implement that. I got the point of that "template" thing, but in any case such thing is totally different thing and shouldn't be handled by posthtml. And that fact that you provide such thing in the loader is wrong path too. It's called "multiple entry points" and in 99% cases isn't good thing in long term.

At very least, "template" is not good/right word for such thing - it more of a "render" or "compile". Actually yea "compile" is the correct term - when opts.template is true, it returns a function that accepts locals.

As about `rollup-plugin-jstransformer`, i just mentioned it, because it not worth to create rollup plugins for each of the hundreds of available engines and transforms out there. That's why such APIs like @jstransformer exist - normalization and stability.

**Comment by michael-ciniawsky** [comment]
> I don't understand you. There is just rollup-plugin-posthtml

I also don't understand what's going on 😛 , `rollup-plugin-posthtml-template` was mentioned here 

> I still can't consider to implement that. I got the point of that "template" thing, but in any case such thing is totally different thing and shouldn't be handled by posthtml. And that fact that you provide such thing in the loader is wrong path too. It's called "multiple entry points" and in 99% cases isn't good thing in long term.

Why? It's just a small opt-in using JS (ECMAScript) Template Literals to interpolate a `{String}` containing HTML markup generated by PostHTML during **runtime** (`fetch()`/AJAX)

At very least, "template" is not good/right word for such thing - it's more of a "render" or "compile". Actually yea "compile" is the correct term - when opts.template is true, it returns a function that accepts locals.

`PostHTML (Parser/Renderer) => Rollup/Webpack => build*( ?runtime** (markup) )`

*build = build time 'compilation'  
**runtime = runtime 'compilation' (optional) 😛

> As about rollup-plugin-jstransformer, i just mentioned it, because it not worth to create rollup plugins for each of the hundreds of available engines and transforms out there. That's why such APIs like @jstransformers exist - normalization and stability.

kk cool, but that's out of scope here :)

**Comment by Shyam-Chen** [comment]
Need to solve this plugin problem (`posthtml-expressions`).

Reference: https://github.com/reshape/expressions

**Comment by tunnckoCore** [comment]
Unsubscribed.
